### PR TITLE
Allow to pass functions when creating ExpiryPolicyBuilder

### PR DESCRIPTION
**Pull request by henri-tremblay**

### CODE REVIEW

File: impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java [hunk 1]
```diff
@@ -64,6 +65,11 @@
     return new TimeToIdleExpiryPolicy(timeToIdle);
   }
 
+  @FunctionalInterface
+  public interface TriFunction<K, V1, V2, R> {
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java]: Could these types be `<T, U, V, R>` to fit with the JDK convention?

File: impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java [hunk 2]
```diff
@@ -75,9 +81,9 @@ public static ExpiryPolicyBuilder expiry() {
     return new ExpiryPolicyBuilder();
   }
 
-  private Duration create = ExpiryPolicy.INFINITE;
-  private Duration access = null;
-  private Duration update = null;
+  private BiFunction<Object, Object, Duration> create = (key, value) -> ExpiryPolicy.INFINITE;
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java]: This type should really have proper generic typing if we're going to expand in to key/value sensitive expiry policies.

File: impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java (hunk 1 above)
Comment (henri-tremblay) [reply impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java]: Yes. I don't know where I took these from. Weird.

File: impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java (hunk 2 above)
Comment (henri-tremblay) [reply impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java]: What do you mean?

Comment (chrisdennis) [reply impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java]: See: https://github.com/henri-tremblay/ehcache3/pull/2

File: impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java [hunk 3]
```diff
@@ -64,36 +65,67 @@
...
   /**
    * Fluent API for creating an {@link ExpiryPolicy} instance where you can specify constant values for creation, access and update time.
    * Unspecified values will be set to {@link ExpiryPolicy#INFINITE INFINITE} for create and {@code null} for access and update, matching
    * the {@link #noExpiration()}  no expiration} expiry.
    *
    * @return an {@link ExpiryPolicy} builder
    */
-  public static ExpiryPolicyBuilder expiry() {
-    return new ExpiryPolicyBuilder();
+  public static <K, V> ExpiryPolicyBuilder<K, V> expiry() {
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java]: I think this is an error in my generics changes... I think this should return `ExpiryPolicyBuilder<Object, Object>`.

File: impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java [hunk 4]
```diff
@@ -64,36 +65,67 @@
...
+    this.create = create;
+    this.access = access;
+    this.update = update;
+  }
 
   /**
    * Set TTL since creation
    *
    * @param create TTL since creation
    * @return this builder
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java]: These `@return` stanzas are now wrong... the builder methods now return new builder instances (the builder has become immutable).

File: impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java [hunk 5]
```diff
@@ -102,12 +134,20 @@ public ExpiryPolicyBuilder create(Duration create) {
...
+    return access((a, b) -> access);
+  }
+
+  /**
+   * Set a function giving the TTL since last access
+   * @param access Function giving the TTL since last access
+   * @return this builder
+   */
+  public <K2 extends K, V2 extends V>ExpiryPolicyBuilder<K2, V2> access(BiFunction<K2, Supplier<? extends V2>, Duration> access) {
+    return new ExpiryPolicyBuilder<>(create, Objects.requireNonNull(access), update);
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java]: Should it be legal to make multiple calls to the builder methods?

File: impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java [hunk 6]
```diff
@@ -116,92 +156,105 @@ public ExpiryPolicyBuilder access(Duration access) {
...
     public int hashCode() {
-      int result = Objects.hashCode(create);
-      result = 31 * result + Objects.hashCode(access);
-      result = 31 * result + Objects.hashCode(update);
-      return result;
+      return ttl.hashCode();
     }
 
     @Override
-    public String toString() {
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java]: We've lost this `toString()` implementation... should we add it back (at least for the TTL and TTI subclasses)?

File: impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java (hunk 3 above)
Comment (henri-tremblay) [reply impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java]: Why?

File: impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java (hunk 6 above)
Comment (henri-tremblay) [reply impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java]: Yes. Why not now that you have added attributes for TTL and TTI.

File: impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java (hunk 5 above)
Comment (chrisdennis) [reply impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java]: Fair enough... might be good to document that the behavior is to over-write... someone might for example think they can specify multiple "access" rules and have the `earliest` expiration time apply.

Comment (henri-tremblay) [reply impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java]: Documented

**Comment by chrisdennis** [comment]
Needs a clean CI build but is otherwise good to go.

**Comment by henri-tremblay** [comment]
@chrisdennis Ok. I triggered a build. Fingers crossed.

**Pull request merged**